src/ucp/core/ucp_worker.c

```diff
@@ -2350,3 +2356,92 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
 
     UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(worker);
 }
+
+static int ucp_worker_keepalive_is_enabled(ucp_worker_h worker)
+{
+    return (worker->context->config.ext.keepalive_timeout != 0.0) &&
+           (worker->context->config.ext.keepalive_num_eps != 0);
+}
+
+static unsigned ucp_worker_keepalive_cb(void *arg)
+{
+    ucp_worker_h worker = (ucp_worker_h)arg;
+    double now          = ucs_time_to_sec(ucs_get_time());
+    unsigned cnt;
+    ucp_ep_ext_gen_t *ep_ext;
+    ucs_list_link_t *iter_begin;
+
+    if (ucs_likely((now - worker->keepalive.last_shot) <
+                   worker->context->config.ext.keepalive_timeout)) {
+        return 0;
+    }
+
+    worker->keepalive.last_shot = now;
+    cnt                         = 0;
+    iter_begin                  = worker->keepalive.iter;
+    /* use own loop for elements because standard for_each skips
+     * head element */
+    /* TODO: use more optimal algo to enumerate EPs to keepalive
+     * (linked list) */
+    do {
+        ep_ext = ucs_list_next(worker->keepalive.iter,
+                               ucp_ep_ext_gen_t, ep_list);
+        worker->keepalive.iter = &ep_ext->ep_list;
+        if (worker->keepalive.iter == &worker->all_eps) {
+            continue;
+        }
+
+        cnt += ucp_ep_do_keepalive(ucp_ep_from_ext_gen(ep_ext));
+        if (cnt >= worker->context->config.ext.keepalive_num_eps) {
+            break;
+        }
+    } while (iter_begin != &ep_ext->ep_list);
+
+    if (cnt == 0) {
+        /* nothing to keepalive? unsubscribe */
+        uct_worker_progress_unregister_safe(worker->uct,
+                                            &worker->keepalive.cb_id);
+    }
+
+    return cnt;
+}
+
+void ucp_worker_keepalive_add_ep(ucp_worker_h worker)
+{
+    if (!ucp_worker_keepalive_is_enabled(worker)) {
+        return;
+    }
+
+    uct_worker_progress_register_safe(worker->uct, ucp_worker_keepalive_cb,
+                                      worker, 0, &worker->keepalive.cb_id);
+}
+
+void ucp_worker_keepalive_remove_ep(ucp_worker_h worker, ucp_ep_h ep)
+{
+    ucp_ep_ext_gen_t *ep_ext;
+
+    if (!ucp_worker_keepalive_is_enabled(worker)) {
+        return;
+    }
+
+    if (worker->keepalive.iter == &ucp_ep_ext_gen(ep)->ep_list) {
+        worker->keepalive.iter = &worker->all_eps;
+    }
+
+    if (worker->keepalive.cb_id == UCS_CALLBACKQ_ID_NULL) {
+        return; /* already unsubscribed */
+    }
+
+    ucs_list_for_each(ep_ext, &worker->all_eps, ep_list) {
+        if (ucp_ep_can_do_keepalive(ucp_ep_from_ext_gen(ep_ext))) {
+            return;
+        }
+    }
```

Comment:
maybe add a counter to worker instead of for_each(all_eps) on every ep_close?

Reply:
will add additional ep list for keepalive EPs

Reply:
it requires extra space in EP, so it can require more changes in EP allocator, so IMO counter is better temporary solution. @yosefe WDYT?

Reply:
here loop looks for at least one EP with keepalive & exits.
added comment